amlip_docs/rst/demo/tensor_inference.rst

```diff
@@ -0,0 +1,189 @@
+    source $CONDA_PREFIX/etc/conda/activate.d/env_vars.sh
+
+Ensure that you have TensorFlow and TensorFlow Hub installed in your Python environment before proceeding.
+You can install them using pip by executing the following commands:
+
+.. code-block:: bash
+
+    pip3 install tensorflow tensorflow-hub tensorflow-object-detection-api nvidia-cudnn-cu11==8.6.0.163
+
+Additionally, it is required to obtain the TensorFlow model from `TensorFlow Hub <https://tfhub.dev/>`_, follow the steps below:
+
+.. code-block:: bash
+
+    cd ~/AML-IP-ws/src/AML-IP/amlip_demo_nodes/amlip_tensorflow_inference_demo/resource/tensorflow/models/
+    wget -O centernet_hourglass_512x512_kpts_1.tar.gz https://tfhub.dev/tensorflow/centernet/hourglass_512x512_kpts/1?tf-hub-format=compressed
+    mkdir centernet_hourglass_512x512_kpts_1
+    tar -xvf centernet_hourglass_512x512_kpts_1.tar.gz -C ./centernet_hourglass_512x512_kpts_1
+
+Run demo
+========
+
+Run Edge Node
+-------------
+
+In the first terminal, run the following command:
+
+.. code-block:: bash
+
+    # Source colcon installation
+    source install/setup.bash
+
+    # To execute Edge Node to send an image to inferred
+    cd ~/AML-IP-ws/src/amlip/amlip_demo_nodes/amlip_tensorflow_inference_demo/amlip_tensorflow_inference_demo
+    python3 edge_node_async.py
+
+Take into account that this node will wait until there is an *Inference Node* running and available
+in the same :term:`LAN` in order to process the inference.
+The expected output is the following:
+
+.. code-block:: bash
+
+    Edge Node AMLEdgeNode.fb.d4.38.13 ready.
+    Edge Node AMLEdgeNode.fb.d4.38.13 sending data.
+
+    Edge Node received inference from AMLInferenceNode.b8.34.4d.a3
+    Edge Node received inference:
+    Box [(0.15590962767601013, 0.21641747653484344), (0.7388607263565063, 0.7326743006706238)] bicycle: 97%
+    Box [(0.16968876123428345, 0.38129815459251404), (0.403958797454834, 0.9422630071640015)] dog: 92%
+    Box [(0.6158109307289124, 0.13117200136184692), (0.9053990244865417, 0.2978983521461487)] truck: 53%
+    Box [(0.6158109307289124, 0.13117200136184692), (0.9053990244865417, 0.2978983521461487)] car: 48%
+    Box [(0.8892407417297363, 0.19558095932006836), (0.933372974395752, 0.2684069573879242)] potted plant: 34%
+    Box [(0.0753115713596344, 0.15651819109916687), (0.13415342569351196, 0.22736744582653046)] motorcycle: 32%
+
+    Edge Node AMLEdgeNode.fb.d4.38.13 closing.
+
+
+Run Inference Node
+------------------
+
+In the second terminal, run the following command to process the inference:
+
+.. code-block:: bash
+
+    # Source colcon installation
+    source install/setup.bash
+
+    # To execute Inference Node with pre-trained model from TensorFlow
+    cd ~/AML-IP-ws/src/AML-IP/amlip_demo_nodes/amlip_tensorflow_inference_demo/amlip_tensorflow_inference_demo
+    python3 inference_node_async.py
```

Comment:
After setting this correctly, executing this line gives the following output and error:

```
$ python3 inference_node_async.py
2023-06-22 09:09:05.735487: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-06-22 09:09:05.772514: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2023-06-22 09:09:05.772870: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-06-22 09:09:06.318896: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
Model Handle at TensorFlow Hub: /home/eprosima/projects/current/annapurna/src/amlip/amlip_demo_nodes/amlip_tensorflow_inference_demo/resource/tensorflow/models/centernet_hourglass_512x512_kpts_1
loading model...
WARNING:absl:Importing a function (__inference_batchnorm_layer_call_and_return_conditional_losses_42408) with ops with unsaved custom gradients. Will likely fail if a gradient is requested.
WARNING:absl:Importing a function (__inference_batchnorm_layer_call_and_return_conditional_losses_209416) with ops with unsaved custom gradients. Will likely fail if a gradient is requested.
...
WARNING:absl:Importing a function (__inference_batchnorm_layer_call_and_return_conditional_losses_55827) with ops with unsaved custom gradients. Will likely fail if a gradient is requested.
WARNING:absl:Importing a function (__inference_batchnorm_layer_call_and_return_conditional_losses_56488) with ops with unsaved custom gradients. Will likely fail if a gradient is requested.
model loaded!
Inference Node AMLInferenceNode.da.86.ed.7b ready. Waiting SIGINT (C^)...
2023-06-22 09:09:34.450904: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'input_tensor' with dtype uint8 and shape [1,?,?,3]
	 [[{{node input_tensor}}]]
2023-06-22 09:09:35.076936: W tensorflow/core/grappler/optimizers/loop_optimizer.cc:907] Skipping loop optimization for Merge node with control input: StatefulPartitionedCall/cond/then/_918/cond/Assert_4/AssertGuard/branch_executed/_1141
2023-06-22 09:09:35.889604: W tensorflow/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 33554432 exceeds 10% of free system memory.
2023-06-22 09:09:35.939295: W tensorflow/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 16777216 exceeds 10% of free system memory.
2023-06-22 09:09:35.939330: W tensorflow/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 34080768 exceeds 10% of free system memory.
2023-06-22 09:09:35.953917: W tensorflow/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 16777216 exceeds 10% of free system memory.
2023-06-22 09:09:35.982900: W tensorflow/tsl/framework/cpu_allocator_impl.cc:83] Allocation of 16777216 exceeds 10% of free system memory.
terminate called after throwing an instance of 'Swig::DirectorMethodException'
  what():  SWIG director method error. In method 'process_inference': AttributeError: module 'tensorflow' has no attribute 'gfile'
Aborted (core dumped)
```